File: 2020/Day3.java
package y_2020;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Day3 {

    final static int n = 323;
    final static int m = 31;
    static char[][] lines = new char[n][m];

    public static void main(String[] args) throws IOException {
        BufferedReader f = new BufferedReader(new FileReader("src/y_2020/day3.in"));

        for (int i = 0; i < n; i++) {
            lines[i] = f.readLine().toCharArray();
        }
        f.close();

        // Part 2
        int[][] increments = {{1,1},{3,1},{5,1},{7,1},{1,2}};

        long product = 1;
        for (int[] in : increments) {
            product *= countTrees(in[0], in[1]);
        }

        System.out.println(product);
    }

    // Part 1
    public static int countTrees(int dx, int dy) {
        int x = 0;
        int y = 0;
        int trees = 0;
        while (y < n) {
            if (lines[y][x] == '#') trees++;
            x += dx;
            if (x >= m) x = x - m;
            y += dy;
        }
        return trees;
    }
}

File: 2015/Day3.py
f = open('day3.in')
s = f.readline()

# Part 1
hm = dict()
hm[(0, 0)] = 1
curr = (0, 0)
for c in s.strip():
    if c == '>':
        curr = (curr[0]+1, curr[1])
    elif c == 'v':
        curr = (curr[0], curr[1]+1)
    elif c == '<':
        curr = (curr[0]-1, curr[1])
    else:
        curr = (curr[0], curr[1]-1)
    
    if curr not in hm:
        hm[curr] = 0
    hm[curr] += 1
print(len(hm))

# Part 2
hm = dict()
hm[(0, 0)] = 1
a = (0, 0)
b = (0, 0)
turn = 0
for c in s.strip():
    curr = a if turn % 2 == 0 else b
    if c == '>':
        curr = (curr[0]+1, curr[1])
    elif c == 'v':
        curr = (curr[0], curr[1]+1)
    elif c == '<':
        curr = (curr[0]-1, curr[1])
    else:
        curr = (curr[0], curr[1]-1)

    if curr not in hm:
        hm[curr] = 0
    hm[curr] += 1
    if turn % 2 == 0:
        a = curr
    else:
        b = curr
    turn += 1
print(len(hm))
